src/matrix.c

```diff
@@ -0,0 +1,225 @@
+}
+
+void matrix_multiply(matrix_t restrict dest, const_matrix_t a, const_matrix_t b,
+                     size_t dim) {
+
+  for (size_t row = 0; row < dim; row++) {
+    for (size_t col = 0; col < dim; col++) {
+      bool value = false;
+      for (size_t k = 0; k < dim; k++) {
+        value ^= matrix_get(a, row, k, dim) & matrix_get(b, k, col, dim);
+      }
+      matrix_set(dest, row, col, value, dim);
+    }
+  }
+}
+
+/**
+ * a step in the gen_nonsing algorithm
+ */
+static void gen_nonsing_dim(matrix_t matrix, const_vector_t b, const_vector_t c,
+                            size_t dim, size_t curr_dim, uint8_t* bmp) {
+  size_t n = dim - curr_dim;
+
+  size_t idx_in_minor = 0; // The index of the curret element in c
+  size_t idx_in_mat =
+      0; // The index corresponding to the place of the place where we have to
+         // put the current element in c in the original matrix
+
+  bool found_nonzero_bit = false; // Found a nonzero bit in c
+
+  for (idx_in_minor = 0; idx_in_minor < n; idx_in_minor++, idx_in_mat++) {
+    while (get_bit(bmp, idx_in_mat)) {
+      idx_in_mat++; // pass all of the selected rows in bmp
+    }
+
+    bool c_value = get_bit(c, idx_in_minor);
+
+    if (c_value) {
+      if (!found_nonzero_bit) {
+        // c_value is the first non-zero calue in c
+        found_nonzero_bit = true;
+        set_bit(bmp, idx_in_mat, true);
+      }
+
+      matrix_elem_add(matrix, idx_in_mat, curr_dim, c_value,
+                      dim); // Add c into the matrix
+      // Add b into the matrix
+      for (size_t v = 0; v < n - 1; v++) {
+        matrix_elem_add(matrix, idx_in_mat, curr_dim + v + 1, get_bit(b, v),
+                        dim);
+      }
+    }
+  }
+}
+
+/**
+ * Iterative algorithm that creates a random nonsingular matrix
+ */
+int matrix_gen_nonsing(matrix_t matrix, size_t dim) {
+  size_t vec_size = round_to_bytes(dim);
+
+  int ret = 0;
+
+  for (size_t i = 0; i < dim; i++) {
+    for (size_t j = 0; j < dim; j++) {
+      matrix_set(matrix, i, j, 0, dim);
+    }
+  }
+
+  uint8_t* bmp = (uint8_t*) calloc(1, vec_size); // Bitmap of minors
+  vector_t b = (vector_t) malloc(vec_size);      // A random vector
+  vector_t c = (vector_t) malloc(vec_size);      // A random non-zero vector
+
+  if (!bmp || !c || !b) {
+    ret = -ENOMEM;
+    goto cleanup;
+  }
+
+  // The iterative algorithm
+  for (size_t j = 0; j < dim; j++) {
+    size_t n = dim - j;
+
+    // generating c: a nonzero coloumn vector
+    ret = gen_nonzero_vector(c, n);
+    if (ret < 0) {
+      goto cleanup;
+    }
+
+    // generating b: a row vector
+    if (!RAND_bytes(b, round_to_bytes(n - 1))) {
+      ret = -EIO;
+      goto cleanup;
+    }
+
+    // placing b and c in the appropriate location in the matrix
+    gen_nonsing_dim(matrix, b, c, dim, j, bmp);
+  }
+
+cleanup:
+  free(bmp);
+  free(b);
+  free(c);
+  return ret;
+}
+
+static void matrix_set_identity(matrix_t matrix, size_t dim) {
+  for (size_t i = 0; i < dim; i++) {
```

Comment:
I think it would probably be better to `memset` the whole thing and set the diagonal.